### ISSUE TITLE
[BUG] cuDF read_parquet occasionally hangs on Maxwell devices

### ISSUE THREAD

**Issue by isolenta** [rapidsai/cudf#3883]
**Describe the bug**
cudf.read_parquet function occasionally hangs on gzip decompression only on Maxwell devices (Volta works fine).

The same issue was initially reported in #3800. Indeed it have beed fixed and doesn't reproduce with test attached in #3800. But even with cuDF fixed with PR #3821 there is another simple test that hangs on Maxwell.

I know that cuDF declares sm_60+, but hanging on Maxwell (sm_52) may indicate some hidden issue.

**Steps/Code to reproduce bug**
```
mkdir data && cd data
wget https://drive.google.com/open?id=1oDz48PRZ3J29Q61bhBcXQ3-Cvo0cC6Lb
cd ..

cat <<EOT >> test.py
import cudf, io, requests, time
from io import StringIO

from os import listdir
from os.path import isfile, join, abspath

mypath = 'data/'

onlyfiles = [f for f in listdir(mypath) if isfile(join(mypath, f))]

t = time.time()

for url in onlyfiles:
    print('Processing file: ', mypath + url)
    tips_df = cudf.read_parquet(mypath + url)
    print(tips_df['column1'])

print('Processing time: ', time.time()-t)
EOT

for x in {1..100}; do python test.py; done
```
cuDF should be built with sm_52 arch.

Script above should hang occasionally on `Start decompress:`  output.

Gdb attached to process shows that process is on cudaStreamSynchronize function:
```
#13 0x00007ff3dfa7561d in cudaStreamSynchronize ()
   from /home/ubuntu/anaconda3/envs/cudf_dev_new/lib/python3.7/site-packages/rmm/_lib/../../../.././libcudart.so.10.0
#14 0x00007ff3477454c0 in cudf::io::parquet::reader::Impl::decompress_page_data(hostdevice_vector<cudf::io::parquet::gpu::ColumnChunkDesc> const&, h

```

**Environment overview (please complete the following information)**
 - Environment location: AWS g3.8xlarge
 - Method of cuDF install: built from source, run with conda

**Environment details**
<details><summary>Click here to see environment details</summary><pre>

     **git***
     Not inside a git repository

     ***OS Information***
     DISTRIB_ID=Ubuntu
     DISTRIB_RELEASE=16.04
     DISTRIB_CODENAME=xenial
     DISTRIB_DESCRIPTION="Ubuntu 16.04.6 LTS"
     NAME="Ubuntu"
     VERSION="16.04.6 LTS (Xenial Xerus)"
     ID=ubuntu
     ID_LIKE=debian
     PRETTY_NAME="Ubuntu 16.04.6 LTS"
     VERSION_ID="16.04"
     HOME_URL="http://www.ubuntu.com/"
     SUPPORT_URL="http://help.ubuntu.com/"
     BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
     VERSION_CODENAME=xenial
     UBUNTU_CODENAME=xenial
     Linux ip-172-31-36-249 4.4.0-1100-aws #111-Ubuntu SMP Wed Dec 4 12:20:15 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

     ***GPU Information***
     Thu Jan 23 15:08:18 2020
     +-----------------------------------------------------------------------------+
     | NVIDIA-SMI 418.87.01    Driver Version: 418.87.01    CUDA Version: 10.1     |
     |-------------------------------+----------------------+----------------------+
     | GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
     | Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
     |===============================+======================+======================|
     |   0  Tesla M60           Off  | 00000000:00:1D.0 Off |                    0 |
     | N/A   39C    P0    47W / 150W |    721MiB /  7618MiB |    100%      Default |
     +-------------------------------+----------------------+----------------------+
     |   1  Tesla M60           Off  | 00000000:00:1E.0 Off |                    0 |
     | N/A   39C    P8    22W / 150W |     11MiB /  7618MiB |      0%      Default |
     +-------------------------------+----------------------+----------------------+

     +-----------------------------------------------------------------------------+
     | Processes:                                                       GPU Memory |
     |  GPU       PID   Type   Process name                             Usage      |
     |=============================================================================|
     |    0      4837      C   python                                       710MiB |
     +-----------------------------------------------------------------------------+

     ***CPU***
     Architecture:          x86_64
     CPU op-mode(s):        32-bit, 64-bit
     Byte Order:            Little Endian
     CPU(s):                32
     On-line CPU(s) list:   0-31
     Thread(s) per core:    2
     Core(s) per socket:    16
     Socket(s):             1
     NUMA node(s):          1
     Vendor ID:             GenuineIntel
     CPU family:            6
     Model:                 79
     Model name:            Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz
     Stepping:              1
     CPU MHz:               2593.070
     CPU max MHz:           3000.0000
     CPU min MHz:           1200.0000
     BogoMIPS:              4600.16
     Hypervisor vendor:     Xen
     Virtualization type:   full
     L1d cache:             32K
     L1i cache:             32K
     L2 cache:              256K
     L3 cache:              46080K
     NUMA node0 CPU(s):     0-31
     Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc aperfmperf pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single kaiser fsgsbase bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx xsaveopt

     ***CMake***
     /home/ubuntu/anaconda3/envs/cudf_dev_new/bin/cmake
     cmake version 3.16.2

     CMake suite maintained and supported by Kitware (kitware.com/cmake).

     ***g++***
     /usr/bin/g++
     g++ (Ubuntu 7.4.0-1ubuntu1~16.04~ppa1) 7.4.0
     Copyright (C) 2017 Free Software Foundation, Inc.
     This is free software; see the source for copying conditions.  There is NO
     warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

     ***nvcc***
     /usr/local/cuda/bin/nvcc
     nvcc: NVIDIA (R) Cuda compiler driver
     Copyright (c) 2005-2018 NVIDIA Corporation
     Built on Sat_Aug_25_21:08:01_CDT_2018
     Cuda compilation tools, release 10.0, V10.0.130

     ***Python***
     /home/ubuntu/anaconda3/envs/cudf_dev_new/bin/python
     Python 3.7.6

     ***Environment Variables***
     PATH                            : /home/ubuntu/anaconda3/envs/cudf_dev_new/bin:/home/ubuntu/anaconda3/condabin:/usr/local/cuda/bin:/home/ubuntu/bin:/home/ubuntu/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
     LD_LIBRARY_PATH                 : /usr/local/cuda/lib64:
     NUMBAPRO_NVVM                   :
     NUMBAPRO_LIBDEVICE              :
     CONDA_PREFIX                    : /home/ubuntu/anaconda3/envs/cudf_dev_new
     PYTHON_PATH                     :

     ***conda packages***
     /home/ubuntu/anaconda3/condabin/conda
     # packages in environment at /home/ubuntu/anaconda3/envs/cudf_dev_new:
     #
     # Name                    Version                   Build  Channel
     _libgcc_mutex             0.1                 conda_forge    conda-forge
     _openmp_mutex             4.5                       0_gnu    conda-forge
     alabaster                 0.7.12                     py_0    conda-forge
     appdirs                   1.4.3                      py_1    conda-forge
     arrow-cpp                 0.15.0           py37h090bef1_2    conda-forge
     aspy.yaml                 1.3.0                      py_0    conda-forge
     attrs                     19.3.0                     py_0    conda-forge
     babel                     2.8.0                      py_0    conda-forge
     backcall                  0.1.0                      py_0    conda-forge
     black                     19.10b0                  py37_0    conda-forge
     bleach                    3.1.0                      py_0    conda-forge
     bokeh                     1.4.0                    py37_0    conda-forge
     boost-cpp                 1.70.0               h8e57a91_2    conda-forge
     brotli                    1.0.7             he1b5a44_1000    conda-forge
     bzip2                     1.0.8                h516909a_2    conda-forge
     c-ares                    1.15.0            h516909a_1001    conda-forge
     ca-certificates           2019.11.28           hecc5488_0    conda-forge
     cached-property           1.5.1                      py_0    conda-forge
     certifi                   2019.11.28               py37_0    conda-forge
     cffi                      1.13.2           py37h8022711_0    conda-forge
     cfgv                      2.0.1                      py_0    conda-forge
     chardet                   3.0.4                 py37_1003    conda-forge
     click                     7.0                      pypi_0    pypi
     cloudpickle               1.2.2                      py_1    conda-forge
     cmake                     3.16.2               h28c56e5_1    conda-forge
     cmake_setuptools          0.1.3                      py_0    rapidsai-nightly
     commonmark                0.9.1                      py_0    conda-forge
     cryptography              2.8              py37h72c5cf5_1    conda-forge
     cudatoolkit               10.0.130                      0    nvidia
     cudf                      0.12.0b0+1748.g837e41a          pypi_0    pypi
     cudnn                     7.6.0                cuda10.0_0    nvidia
     cupy                      7.0.0            py37he57b8b9_2    conda-forge
     cython                    0.29.14          py37he1b5a44_0    conda-forge
     cytoolz                   0.10.1           py37h516909a_0    conda-forge
     dask                      2.9.2+9.ge49cf4d          pypi_0    pypi
     dask-cudf                 0.12.0b0+1748.g837e41a          pypi_0    pypi
     decorator                 4.4.1                      py_0    conda-forge
     defusedxml                0.6.0                      py_0    conda-forge
     distributed               2.9.3+5.ga891a83          pypi_0    pypi
     dlpack                    0.2                  he1b5a44_1    conda-forge
     docutils                  0.16                     py37_0    conda-forge
     double-conversion         3.1.5                he1b5a44_2    conda-forge
     editdistance              0.5.3            py37he1b5a44_0    conda-forge
     entrypoints               0.3                   py37_1000    conda-forge
     expat                     2.2.5             he1b5a44_1004    conda-forge
     fastavro                  0.22.9           py37h516909a_0    conda-forge
     fastrlock                 0.4             py37he1b5a44_1000    conda-forge
     flake8                    3.7.9                    py37_0    conda-forge
     flatbuffers               1.11.0               he1b5a44_0    conda-forge
     freetype                  2.10.0               he983fc9_1    conda-forge
     fsspec                    0.6.2                      py_0    conda-forge
     future                    0.18.2                   py37_0    conda-forge
     gflags                    2.2.2             he1b5a44_1002    conda-forge
     glog                      0.4.0                he1b5a44_1    conda-forge
     gmp                       6.2.0                he1b5a44_0    conda-forge
     grpc-cpp                  1.23.0               h18db393_0    conda-forge
     heapdict                  1.0.1                      py_0    conda-forge
     hypothesis                5.3.0                      py_0    conda-forge
     icu                       64.2                 he1b5a44_1    conda-forge
     identify                  1.4.10                     py_0    conda-forge
     idna                      2.8                   py37_1000    conda-forge
     imagesize                 1.2.0                      py_0    conda-forge
     importlib_metadata        1.4.0                    py37_0    conda-forge
     ipykernel                 5.1.3            py37h5ca1d4c_0    conda-forge
     ipython                   7.11.1           py37h5ca1d4c_0    conda-forge
     ipython_genutils          0.2.0                      py_1    conda-forge
     isort                     4.3.21                   py37_0    conda-forge
     jedi                      0.15.2                   py37_0    conda-forge
     jinja2                    2.10.3                     py_0    conda-forge
     jpeg                      9c                h14c3975_1001    conda-forge
     jsonschema                3.2.0                    py37_0    conda-forge
     jupyter_client            5.3.4                    py37_1    conda-forge
     jupyter_core              4.6.1                    py37_0    conda-forge
     krb5                      1.16.4               h2fd8d38_0    conda-forge
     ld_impl_linux-64          2.33.1               h53a641e_7    conda-forge
     libblas                   3.8.0               14_openblas    conda-forge
     libcblas                  3.8.0               14_openblas    conda-forge
     libcurl                   7.65.3               hda55be3_0    conda-forge
     libedit                   3.1.20170329      hf8c457e_1001    conda-forge
     libevent                  2.1.10               h72c5cf5_0    conda-forge
     libffi                    3.2.1             he1b5a44_1006    conda-forge
     libgcc-ng                 9.2.0                h24d8f2e_2    conda-forge
     libgfortran-ng            7.3.0                hdf63c60_4    conda-forge
     libgomp                   9.2.0                h24d8f2e_2    conda-forge
     liblapack                 3.8.0               14_openblas    conda-forge
     libllvm8                  8.0.1                hc9558a2_0    conda-forge
     libopenblas               0.3.7                h5ec1e0e_6    conda-forge
     libpng                    1.6.37               hed695b0_0    conda-forge
     libprotobuf               3.8.0                h8b12597_0    conda-forge
     librmm                    0.12.0a200122      cuda10.0_194    rapidsai-nightly
     libsodium                 1.0.17               h516909a_0    conda-forge
     libssh2                   1.8.2                h22169c7_2    conda-forge
     libstdcxx-ng              9.2.0                hdf63c60_2    conda-forge
     libtiff                   4.1.0                hfc65ed5_0    conda-forge
     libuv                     1.34.0               h516909a_0    conda-forge
     llvmlite                  0.31.0           py37h8b12597_0    conda-forge
     locket                    0.2.0                      py_2    conda-forge
     lz4-c                     1.8.3             he1b5a44_1001    conda-forge
     markdown                  3.0.1                    pypi_0    pypi
     markupsafe                1.1.1            py37h516909a_0    conda-forge
     mccabe                    0.6.1                      py_1    conda-forge
     mistune                   0.8.4           py37h516909a_1000    conda-forge
     more-itertools            8.1.0                      py_0    conda-forge
     msgpack-python            0.6.2            py37hc9558a2_0    conda-forge
     mypy_extensions           0.4.3                    py37_0    conda-forge
     nbconvert                 5.6.1                    py37_0    conda-forge
     nbformat                  5.0.3                      py_0    conda-forge
     nbsphinx                  0.5.0                      py_0    conda-forge
     nccl                      2.4.6.1              cuda10.0_0    nvidia
     ncurses                   6.1               hf484d3e_1002    conda-forge
     nodeenv                   1.3.4                      py_0    conda-forge
     notebook                  6.0.3                    py37_0    conda-forge
     numba                     0.47.0           py37hb3f55d8_0    conda-forge
     numpy                     1.17.5           py37h95a1406_0    conda-forge
     numpydoc                  0.9.2                      py_0    conda-forge
     nvstrings-cudaunknown     0.0.0.dev0               pypi_0    pypi
     olefile                   0.46                       py_0    conda-forge
     openssl                   1.1.1d               h516909a_0    conda-forge
     packaging                 20.0                       py_0    conda-forge
     pandas                    0.25.3           py37hb3f55d8_0    conda-forge
     pandoc                    1.19.2                        0    conda-forge
     pandocfilters             1.4.2                      py_1    conda-forge
     parquet-cpp               1.5.1                         2    conda-forge
     parso                     0.5.2                      py_0    conda-forge
     partd                     1.1.0                      py_0    conda-forge
     pathspec                  0.7.0                      py_0    conda-forge
     pexpect                   4.7.0                    py37_0    conda-forge
     pickleshare               0.7.5                 py37_1000    conda-forge
     pillow                    7.0.0            py37hefe7db6_0    conda-forge
     pip                       19.3.1                   py37_0    conda-forge
     pluggy                    0.13.0                   py37_0    conda-forge
     pre_commit                1.18.1                   py37_0    conda-forge
     prometheus_client         0.7.1                      py_0    conda-forge
     prompt_toolkit            3.0.2                      py_0    conda-forge
     psutil                    5.6.7            py37h516909a_0    conda-forge
     ptyprocess                0.6.0                   py_1001    conda-forge
     py                        1.8.1                      py_0    conda-forge
     pyarrow                   0.15.0           py37h8b68381_1    conda-forge
     pycodestyle               2.5.0                      py_0    conda-forge
     pycparser                 2.19                     py37_1    conda-forge
     pyflakes                  2.1.1                      py_0    conda-forge
     pygments                  2.5.2                      py_0    conda-forge
     pyopenssl                 19.1.0                   py37_0    conda-forge
     pyparsing                 2.4.6                      py_0    conda-forge
     pyrsistent                0.15.7           py37h516909a_0    conda-forge
     pysocks                   1.7.1                    py37_0    conda-forge
     pytest                    5.3.4                    py37_0    conda-forge
     python                    3.7.6                h357f687_2    conda-forge
     python-dateutil           2.8.1                      py_0    conda-forge
     pytz                      2019.3                     py_0    conda-forge
     pyyaml                    5.3              py37h516909a_0    conda-forge
     pyzmq                     18.1.1           py37h1768529_0    conda-forge
     rapidjson                 1.1.0             he1b5a44_1002    conda-forge
     re2                       2020.01.01           he1b5a44_0    conda-forge
     readline                  8.0                  hf8c457e_0    conda-forge
     recommonmark              0.6.0                      py_0    conda-forge
     regex                     2020.1.8         py37h516909a_0    conda-forge
     requests                  2.22.0                   py37_1    conda-forge
     rhash                     1.3.6             h14c3975_1001    conda-forge
     rmm                       0.12.0a200122          py37_194    rapidsai-nightly
     send2trash                1.5.0                      py_0    conda-forge
     setuptools                45.1.0                   py37_0    conda-forge
     six                       1.14.0                   py37_0    conda-forge
     snappy                    1.1.7             he1b5a44_1003    conda-forge
     snowballstemmer           2.0.0                      py_0    conda-forge
     sortedcontainers          2.1.0                      py_0    conda-forge
     sphinx                    2.3.1                      py_0    conda-forge
     sphinx-markdown-tables    0.0.12                   pypi_0    pypi
     sphinx_rtd_theme          0.4.3                      py_0    conda-forge
     sphinxcontrib-applehelp   1.0.1                      py_0    conda-forge
     sphinxcontrib-devhelp     1.0.1                      py_0    conda-forge
     sphinxcontrib-htmlhelp    1.0.2                      py_0    conda-forge
     sphinxcontrib-jsmath      1.0.1                      py_0    conda-forge
     sphinxcontrib-qthelp      1.0.2                      py_0    conda-forge
     sphinxcontrib-serializinghtml 1.1.3                      py_0    conda-forge
     sphinxcontrib-websupport  1.1.2                      py_0    conda-forge
     sqlite                    3.30.1               hcee41ef_0    conda-forge
     streamz                   0.5.2                    pypi_0    pypi
     tblib                     1.6.0                      py_0    conda-forge
     terminado                 0.8.3                    py37_0    conda-forge
     testpath                  0.4.4                      py_0    conda-forge
     thrift-cpp                0.12.0            hf3afdfd_1004    conda-forge
     tk                        8.6.10               hed695b0_0    conda-forge
     toml                      0.10.0                     py_0    conda-forge
     toolz                     0.10.0                     py_0    conda-forge
     tornado                   6.0.3            py37h516909a_0    conda-forge
     traitlets                 4.3.3                    py37_0    conda-forge
     typed-ast                 1.4.1            py37h516909a_0    conda-forge
     typing_extensions         3.7.4.1                  py37_0    conda-forge
     uriparser                 0.9.3                he1b5a44_1    conda-forge
     urllib3                   1.25.7                   py37_0    conda-forge
     virtualenv                16.7.5                     py_0    conda-forge
     wcwidth                   0.1.8                      py_0    conda-forge
     webencodings              0.5.1                      py_1    conda-forge
     wheel                     0.33.6                   py37_0    conda-forge
     xz                        5.2.4             h14c3975_1001    conda-forge
     yaml                      0.2.2                h516909a_1    conda-forge
     zeromq                    4.3.2                he1b5a44_2    conda-forge
     zict                      1.0.0                    pypi_0    pypi
     zipp                      2.0.0                      py_1    conda-forge
     zlib                      1.2.11            h516909a_1006    conda-forge
     zstd                      1.4.3                h3b9ef0a_0    conda-forge

</pre></details>

**Comment by kkraus14** [comment]
@isolenta `cuda-memcheck` doesn't report any invalid read/writes when your reproducer above is run under it?

**Comment by isolenta** [comment]
@kkraus14 `cuda-memcheck` doesn't report anything

**Comment by OlivierNV** [comment]
@isolenta Can you confirm that it only hangs when using gzip-compressed parquet files (as opposed to snappy compression) ?
I can prepare an instrumented version of the decompressor kernel to timeout gracefully and dump the decompressor state in the reserved and/or status members of inflate_status_s to help debugging this.

**Comment by isolenta** [comment]
@OlivierNV I'm confirming: the same parquet data but compressed with snappy doesn't hang with the same test. Only gzip decompression is affected.

Instrumented version would help to figure out. I'll run and collect all debugging information then.

**Comment by OlivierNV** [comment]
@isolenta Could you try if PR #4369 resolves this issue ?

**Comment by isolenta** [comment]
@OlivierNV indeed, PR @4369 fixes this issue. Thanks!

**Comment by kkraus14** [comment]
Fixed by #4369
